pkg/server/container_create.go

```diff
@@ -134,26 +134,35 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
 	logrus.Debugf("Use OCI %+v for container %q", ociRuntime, id)
 
 	// Create container root directory.
-	containerRootDir := getContainerRootDir(c.config.RootDir, id)
+	containerRootDir := getContainerRootDir(c.config.StateDir, id)
 	if err = c.os.MkdirAll(containerRootDir, 0755); err != nil {
 		return nil, errors.Wrapf(err, "failed to create container root directory %q",
 			containerRootDir)
 	}
+	persistentContainerRootDir := getContainerRootDir(c.config.RootDir, id)
+	if err = c.os.MkdirAll(persistentContainerRootDir, 0755); err != nil {
+		return nil, errors.Wrapf(err, "failed to create persistent container root directory %q",
+			persistentContainerRootDir)
+	}
 	defer func() {
 		if retErr != nil {
-			// Cleanup the container root directory.
+			// Cleanup the container root directories.
 			if err = c.os.RemoveAll(containerRootDir); err != nil {
 				logrus.WithError(err).Errorf("Failed to remove container root directory %q",
 					containerRootDir)
 			}
+			if err = c.os.RemoveAll(persistentContainerRootDir); err != nil {
```

Comment:
Each defer cleanup should be right after the creation.
e.g. `containerRootDir` cleanup won't run when `c.os.MkdirAll(persistentContainerRootDir, 0755)` returns error, which is not what we want.